Use correct sample bit width

viinikv opened the pull request:
snd_pcm_hw_params_get_sbits gives the number of significant bits, not the actual number of bits stored. Change to snd_pcm_format_physical_width.

This fixes a bug where, for example on my hardware:
format = 'S32_LE'
significant bits = 24
physical bits = 32

the program will segfault because the allocated buffer is too small.

RonaldAJ commented:
I had to check the definition of significant, nominal and physical:

`PCM formats

The full list of formats present the [snd_pcm_format_t](https://www.alsa-project.org/alsa-doc/alsa-lib/group___p_c_m.html#gaa14b7f26877a812acbb39811364177f8) type. The 24-bit linear samples use 32-bit physical space, but the sample is stored in the lower three bytes. Some hardware does not support processing of full range, thus you may get the significant bits for linear samples via [snd_pcm_hw_params_get_sbits()](https://www.alsa-project.org/alsa-doc/alsa-lib/group___p_c_m___h_w___params.html#ga8bb1180b62a4225f0e094b95ac5d3674) function. The example: ICE1712 chips support 32-bit sample processing, but low byte is ignored (playback) or zero (capture). The function [snd_pcm_hw_params_get_sbits()](https://www.alsa-project.org/alsa-doc/alsa-lib/group___p_c_m___h_w___params.html#ga8bb1180b62a4225f0e094b95ac5d3674) returns 24 in this case.`

So the proposed change to the calculation of self->framesize is justified.

RonaldAJ commented:
For completeness it might be desirable to also return the nominal bits in info() and dumpinfo(). That is the output of snd_pcm_format_width().  This would be redundant with some of the format information, but would not require string processing before use.

ossilator commented:
good catch!
pedantically, the one-liner bug fix should be split off the addition of the new info.

ossilator commented:
i'm confused about nominal vs. significant bits. i presume one defines the range (i.e., LSBs that can be non-zero) and the other the precision (i.e., MSBs that are not ignored), but alsa documentation being what it is, i haven't found the authoritative definitions yet. suppose a really stupid format which stores 16 bits in the middle of a 32-bit word - what would be numbers be?

larsimmisch reviewed (approved):
Thanks! Looks good to me.

viinikv commented:
@ossilator I removed the new "physical_bits" info for now. (I changed the original 54cb698 commit, hopefully nobody has it in their tree)

I think the bit widths are:
```
MSB |0000 XXXX XXXX 0000| LSB
    |-significant--|
         |---nominal----|
    |-----physical------|
```

For a reference, for nominal bits, maybe the usage in alsa-lib [/test/pcm.c](https://www.alsa-project.org/alsa-doc/alsa-lib/_2test_2pcm_8c-example.html#:~:text=int%20format_bits)
```c
    int format_bits = snd_pcm_format_width(format);
    unsigned int maxval = (1 << (format_bits - 1)) - 1;
```

And snd_pcm_hw_params_get_sbits returns msbits, which I think is the same as in the kernel, which is documented in include/uapi/sound/asound.h
```c
unsigned int msbits;            /* R: used most significant bits */
```
and
```c
/*      
 * For S32/U32 formats, 'msbits' hardware parameter is often used to deliver information about the
 * available bit count in most significant bit. It's for the case of so-called 'left-justified' or
 * `right-padding` sample which has less width than 32 bit.
 */     
```

Maybe some documentation of these could be added to doc/terminology.rst. Shoud I add physical_bits and nominal_bits info and documentation here or as a separate pull request or does some of you prefer to do that yourself?

ossilator commented:
cool, thanks for the thorough investigation! :+1: 
you're most welcome to implement these changes yourself; i think neither of us here has too little to do otherwise. :joy: 
a "sample size" section in the terminology document sounds appropriate, but make sure that it's linked from the relevant info() entries in the table.

the forced push is just fine, and it's in fact how i _expect_ things to be done.
you can put all these changes in the same PR if it's convenient for you - i care only that the commits are atomic.

larsimmisch commented:
First of all: thanks a lot. I like the ascii art in this thread and I'd love to see it in the documentation. Whether in this PR or a new one is up to you.

@ossilator, `Rebase and marge would be most appropriate here (assuming, say, a doc update commit was pushed to this branch)?

@viinikv out of curiosity: is your `Rebase and merge` button green?

ossilator commented:
> Rebase and merge would be most appropriate here (assuming, say, a doc update commit was pushed to this branch)?

yes.
let's consider the options:
- a squash merge destroys the history. it makes sense if the history is an utter mess, so it's not a relevant option when our policy is to insist on PRs with a clean history.
- a merge commit preserves the PR history and creates a "delimiter" in the mainline. one may choose this as a policy, but personally i think that it just complicates the history for no good reason.
- rebase preserves the PR history and keeps the mainline history linear, and thus easy to follow. also, when no concurrent contributions are made, the rebase is empty, so the result is mostly as if a direct push had been done, which seems most intuitive to me.
hence it's my preference, and is universally applicable unless the PR contains a merge that needs to be preserved (which i don't expect to happen in this project).
